### PR TITLE
Extend the validation function

### PR DESCRIPTION
**Pull request by andrey-churkin**
### Changes

- The validation function can return metrics/logits for each data item from a dataset now.

### Reason for changes

- Improve execution time for algorithm

### Related tickets

N/A

### Tests

N/A

### CODE REVIEW

**Review by alexsu52 — COMMENTED** [review]
Great proposal. Please, update examples as well.

File: nncf/quantization/algorithms/accuracy_control/evaluator.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        """
+        Determines mode based on the type of returned value from the
+        validation function.
+        :param model_for_inference: Model to validate.
+        :param dataset: Dataset to validate the model.
+        """
+        data_item = dataset.get_data([0])
+        # pylint: disable=W0703
+        try:
+            _, values_for_each_item = self._validation_fn(model_for_inference, data_item)
```
[the remark this answers is not in the document]
Comment (andrey-churkin) [reply nncf/quantization/algorithms/accuracy_control/evaluator.py:109]: We assume `validation_fn()` should not fail when we want to calculate a metric for the full dataset. This `try-except` block is required to figure out if we can calculate a metric for a single sample or not. The following cases are possible
- `validation_fn()` failed for a single sample then `metric_mode=False`
- `validation_fn()` did not fail for a single sample and `values_for_each_item is None` then `metric_mode=True`
- `validation_fn()` did not fail for a single sample and `values_for_each_item[0] is float`  then `metric_mode=True`
- `validation_fn()` did not fail for a single sample and `values_for_each_item[0] is Output` then `metric_mode=False`

**Comment by andrey-churkin** [comment]
> Please, update examples as well.

What do you mean? Do you want to use the native implementation in examples?

File: nncf/experimental/openvino/quantization/quantize.py [hunk 2]
```diff
@@ -88,26 +84,9 @@ def quantize_with_accuracy_control(model: ov.Model,
...
-    nncf_logger.info(f'Metric of quantized model: {quantized_metric}')
-
-    accuracy_aware_loop = QuantizationAccuracyRestorer(max_num_iterations=max_num_iterations,
-                                                       max_drop=max_drop)
-    quantized_model = accuracy_aware_loop.restore_accuracy(model, initial_metric,
-                                                           quantized_model, quantized_metric,
-                                                           validation_dataset, validation_fn)
+    accuracy_restorer = QuantizationAccuracyRestorer(max_num_iterations=max_num_iterations,
+                                                     max_drop=max_drop)
+    quantized_model = accuracy_restorer.apply(model, quantized_model, validation_dataset, validation_fn)
```
[the remark this answers is not in the document]
Comment (andrey-churkin) [reply nncf/experimental/openvino/quantization/quantize.py:93]: Done

File: nncf/quantization/algorithms/accuracy_control/evaluator.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+from nncf.data.dataset import Dataset
+from nncf.common.factory import EngineFactory
+from nncf.quantization.algorithms.accuracy_control.backend import AccuracyControlAlgoBackend
+
+
+TModel = TypeVar('TModel')
+
+
+class Output:
```
[the remark this answers is not in the document]
Comment (andrey-churkin) [reply nncf/quantization/algorithms/accuracy_control/evaluator.py:24]: Something like this
```python
def validate(compiled_model: ov.CompiledModel,
             val_loader: torch.utils.data.DataLoader) -> Tuple[float, List[Output]]:
    predictions = []
    references = []
    logits_for_each_sample = []

    compiled_model = ov.compile_model(model)
    output = compiled_model.outputs[0]

    for images, target in tqdm(val_loader):
        pred = compiled_model(images)[output]
        logits = Output()
        logits.register(pred)
        logits_for_each_sample.append(logits)
        predictions.append(np.argmax(pred, axis=1))
        references.append(target)

    predictions = np.concatenate(predictions, axis=0)
    references = np.concatenate(references, axis=0)
    return accuracy_score(predictions, references), logits_for_each_sample 
```

File: nncf/quantization/algorithms/accuracy_control/evaluator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexsu52) [reply nncf/quantization/algorithms/accuracy_control/evaluator.py:109]: validation_fn() did not fail for a single sample, `metric_value is None` and `values_for_each_item is None` then `metric_mode=?`

[the remark this answers is not in the document]
Comment (andrey-churkin) [reply nncf/quantization/algorithms/accuracy_control/evaluator.py:109]: The metric_value can not be `None`. It is not a valid case. Please look at the return value type:
```
Tuple[float, Union[List[float], List[Output], None]]]
```
The first element of tuple is `float`, not `Optional[float]`

File: nncf/quantization/algorithms/accuracy_control/evaluator.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexsu52) [reply nncf/quantization/algorithms/accuracy_control/evaluator.py:24]: Thanks for the reply! As far as I understand, the goal is to introduce an interface that will allow you to distinguish whether the float metric for each sample and or the output tensors of the model, which will later be used to calculate the MSE.

```
def validate(compiled_model: ov.CompiledModel,
             val_loader: torch.utils.data.DataLoader) -> Tuple[float, List[Output]]:
    predictions = []
    references = []
    model_output_for_each_sample = []

    compiled_model = ov.compile_model(model)
    output = compiled_model.outputs[0]

    for images, target in tqdm(val_loader):
        pred = compiled_model(images)[output]
        output_tensors = [nncf.Tensor(pred)]
        model_output_for_each_sample.append(output_tensors)
        predictions.append(np.argmax(pred, axis=1))
        references.append(target)

    predictions = np.concatenate(predictions, axis=0)
    references = np.concatenate(references, axis=0)
    return accuracy_score(predictions, references), model_output_for_each_sample
```
What do you think about it?

**Comment by alexsu52** [comment]
> > Please, update examples as well.
> 
> What do you mean? Do you want to use the native implementation in examples?

I meant that the examples should be adjusted to changes.

File: nncf/quantization/algorithms/accuracy_control/evaluator.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrey-churkin) [reply nncf/quantization/algorithms/accuracy_control/evaluator.py:24]: We can introduce the `nncf.Tensor` class and use it. The return type will be
```python
Tuple[float, Union[List[float], List[List[nncf.Tensor]], None]
```

File: nncf/quantization/algorithms/accuracy_control/evaluator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (alexsu52) [reply nncf/quantization/algorithms/accuracy_control/evaluator.py:109]: Okay, what happens if `None` will come?

**Review by alexsu52 — COMMENTED** [review]
Please, bring this PR on the offline review.

**Comment by andrey-churkin** [comment]
Reopened here: https://github.com/openvinotoolkit/nncf/pull/1809

File: nncf/quantization/algorithms/accuracy_control/evaluator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrey-churkin) [reply nncf/quantization/algorithms/accuracy_control/evaluator.py:109]: Fixed
